More cslib_array functions and friends

CareyJWilliams opened the pull request:
Six commits:
- removing excess braces
- cleaning up cslib_string's param info
- adding a bunch more functions to cslib_array
- Add filter support back to cslib_menu (via routines)
- Fix a bug where viewing logs still used the original (removed) cslib_menu filter
- Add 'array_ref' parameter to passed array routines

CareyJWilliams committed:
Remove extraneous curly braces

CareyJWilliams committed:
Tidy up parameter information in cslib_string

There were some mistakes and lazy descriptions.
Resolves #15.

CareyJWilliams committed:
Further additions to cslib_array

Adds the following to cslib_array:
- reverse (reverses the array in-place)
- for_each (calls a routine for each i)
- test_all (checks ALL pass routine)
- test_any (checks if ANY pass routine)
- join (joins an array's items into a string)
- set (set array contents and len via params)
- set_from (set but from a specified index)

danspinola commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
 *return
 
 
+*comment TO_STRING
```
I suggest we use another name. Even in serious programming languages "to\_string" is not used to mean joining elements in an array. At least I haven't seen it. Maybe "join"?

CareyJWilliams replied:
I agree! "Join" would have been my first choice, but I'm trying to think as if I'm not familiar with programming. What is the most intuitive name we can give it?

I'm not married to "to string" but maybe join isn't the best? Maybe one to discuss on discord.

danspinola commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*comment        p_array_ref(string): the name of the array to convert to a string
+*comment        p_separator(string)?: the string to insert between each array element
+*comment    returns:
+*comment        (any): all array elements as a string
```
Well, technically, the return is a string. 😂

CareyJWilliams replied:
Correct, this should be a -2!
Good catch ;)

danspinola commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*return
+
+
+*comment TEST_SOME
```
Maybe "test\_any"?

CareyJWilliams replied:
I can go with that!
On the topic: test or check?

danspinola replied:
🤔

danspinola replied:
test 😂

danspinola reviewed (approved):
Looks great! I've made some suggestions, but merely about nomenclature. Nothing impeding.

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -155,7 +155,7 @@
 *comment        p_value(any): the value to be added to the end of the array
 *label push
 *params p_array_ref p_value
-*gosub insert p_array_ref p_value ({{p_array_ref&"_count"}} + 1)
```
maybe we could make it uniform across functions, so that the prefix includes the underscore (for stuff like {prefix & x} )

CareyJWilliams replied:
That'd be putting the onus on the user. Yes, it makes our code the tiniest bit easier to read, but we're trying to accommodate end-users who don't have a programming background here -- I think that would be a move in the wrong direction.

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*comment ------------------------------
+*comment    params:
+*comment        p_array_ref(string): the name of the array to convert to a string
+*comment        p_separator(string)?: the string to insert between each array element
```
does the question mark means optional? if so, please specify

CareyJWilliams replied:
Yeah, I was trying to think a non-verbose way of indicating that. I think I've used '?' in another spot or two. I can add the words optional to the description as well.

danspinola replied:
Maybe it would be good to settle on some general conventions and include an example in the readme explaining how to read the docs.

I like not being verbose by the way 😂

CareyJWilliams replied:
I think when we review the contribution guide, we could cover that. Probably be worth having a style guide as well.

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*temp n 1
+*label _to_string_loop
+*if (n < count)
+    *set cslib_ret &({p_array_ref&"_${n}"}&p_separator)
```
if you have underscore in the prefix, it's MUCH easier to read
{prefix & n }

CareyJWilliams replied:
See above

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*label _reverse_loop
+*if (s_idx < e_idx)
+    *set store {p_array_ref&"_${s_idx}"}
+    *set {p_array_ref&"_${s_idx}"} {p_array_ref&"_${e_idx}"}
```
same here {prefix & id}

CareyJWilliams replied:
Above

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*if (s_idx < e_idx)
+    *set store {p_array_ref&"_${s_idx}"}
+    *set {p_array_ref&"_${s_idx}"} {p_array_ref&"_${e_idx}"}
+    *set {p_array_ref&"_${e_idx}"} store
```
A trick to swap two values. The *gosub swap may be more readable

https://forum.choiceofgames.com/t/new-choicescript-features-implicit-control-flow-gosub-parameters/30042/24

CareyJWilliams replied:
It's a neat trick, but I don't think it will improve things here as we have to set the global values (not just the temps). You'll also trash p_array_ref on a second call to *params.

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*return
+
+
+*comment REVERSE
```
nice!

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*comment
+*comment And should return a new value for the element.
+*comment E.g.:
+*comment    *label capitalize_str
```
can you add the "foreach" call to the example?

CareyJWilliams replied:
Do you mean the Examples folder example? As in char_creator? I plan to follow up with more patches to adjust that/add others at a later date.

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*temp n 1
+*label _for_each_loop
+*if (n <= count)
+    *gosub_scene {p_scene} {p_label} {p_array_ref&"_${n}"} n p_array_ref
```
wait, why do you need to pass
array[n]   n   array
can't the function determine array[n] out of array and n?

CareyJWilliams replied:
Yes, but you've got multiple comments on this patch saying how ugly doing that is... ;)
Best we keep that away from the end-user, so this is a convenience value (it doesn't cost us anything).

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*comment And should return a new value for the element.
+*comment E.g.:
+*comment    *label capitalize_str
+*comment    *params value index array_ref
```
I believe 'value' is redundant. The function can just do array[n]

CareyJWilliams replied:
No it can't, because it's a reference. It'd have to do `{p_array_ref&"_${n}"}` (see above).

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*comment
+*comment The given routine should expect
+*comment to be passed three parameters:
+*comment - The element value
```
unneeded

CareyJWilliams replied:
Above.

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*comment And should return a boolean (true = condition met).
+*comment E.g.:
+*comment    *label is_short_string
+*comment    *params value index
```
the example is misleading as it has only two args (above it says it should have three)

CareyJWilliams replied:
Good spot, I'll add that.

diegoboston commented on scenes/cslib_array.txt:
```diff
@@ -247,6 +247,180 @@
+*comment
+*comment The given routine should expect
+*comment to be passed three parameters:
+*comment - The element value
```
same

diegoboston commented on scenes/cslib_string.txt:
```diff
@@ -39,8 +39,9 @@
 *comment a given seperator.
```
typo: sepArator

diegoboston commented on scenes/cslib_string.txt:
```diff
@@ -161,10 +161,10 @@
-*comment		p_old(string): the old substring to match
-*comment		p_new(string): the new substring to replace those matches
+*comment		p_old_str(string): the string value to replace
+*comment		p_new_str(string): the string value to replace instances of p_old_str with
```
"to use as a replacement"

diegoboston commented on scenes/cslib_string.txt:
```diff
@@ -203,10 +203,10 @@
 *comment 	params:
-*comment		p_str(string): a string to search for p_char
-*comment		p_char(string): a string of length 1
+*comment		p_str(string): a string in which to search for p_char
```
consider: in which -> where

CareyJWilliams replied:
Hmm... That also doesn't sound quite right to me. What about just "_a string to search for p_char_"?

diegoboston commented on scenes/test_array.txt:
```diff
@@ -241,6 +241,239 @@
+*set test_arr_2 "jane"
+*set test_arr_count 2
+*gosub_scene cslib_array for_each "test_arr" "test_array" "for_each_capitalize"
+*if (false)
```
that's a nice trick!! You could probably even do something like

*if (false and "Define a function")
*if ("Define a function" == "")

This is better than what my compiler does now:

*goto skip_xxx
*label xxx
FUNCTION BODY
*label skip_xxx

danspinola replied:
I've seen GoldenSilver use it in his game files. (I think)

CareyJWilliams replied:
I'm not sure I follow with the *if(false and "Define a function") stuff, but I'm curious?

diegoboston commented on scenes/test_array.txt:
```diff
@@ -241,6 +241,239 @@
+
+*gosub test_start "for_each #1 (empty)"
+*set test_arr_count 0
+*comment basically just test this doesn't crash
```
respect!

diegoboston commented on scenes/test_array.txt:
```diff
@@ -241,6 +241,239 @@
+	*comment check every 2nd element is identical
+	*comment to the one before it
+	*label test_all_duplicate_line
+	*params value index arr_ref
```
Value is unused, as I mentioned above.

diegoboston commented on scenes/test_array.txt:
```diff
@@ -241,6 +241,239 @@
+*gosub_scene cslib_array test_some "test_arr" "test_array" "is_palindrome"
+*if (false)
+	*comment check if a word spells itself backwards
+	*label is_palindrome
```
do you want to move this to the string module? It's a cute example and people may use it for a magic incantation.
*text_input password
*if is_palindrome...  (not real CS)

CareyJWilliams replied:
So for the "complex" examples, I very nearly did call some cslib_string routines, but then I remembered what you said about not having tests rely on the functionality of other modules (which was a good point).

The palindrome routine is fun, but I don't know if it's useful enough to warrant a place in cslib_string itself.
It's kind of a slippery slope to having all sorts of not-very useful (day-to-day) stuff to maintain and run tests on.

danspinola replied:
good point

diegoboston reviewed (approved):
Great! A few minor suggestions.

CareyJWilliams committed:
[bug fix] visit_logs should use the new menu filter

When removing the original cslib_menu build_array_filter
implementation this was missed. This updates it to use
the newly added routine variant.

CareyJWilliams committed:
Return 'filter' support to cslib_array via routines

We removed the filter support from cslib_menu to
simplify the module, thinking we could rely on
the 'filter' routine from cslib_array.
Unfortunately that was an oversight. When you use
filter in cslib_array, you're creating a new array,
with new element indexes.

This means that the specific use-case of using cslib_menu's
return to index into another attribute of the array is lost.

E.g. with cslib's test suite:
*gosub_scene cslib_menu build_array_filter "module" ...
${module_log[cslib_ret}

The above is no longer possible with cslib_array filter.

This patch returns that functionality to cslib_menu via
a routine implementation (passing a scene+label), which
has a few advantages:
- It matches up with cslib_array implementations
- It simplifies the code in cslib_menu (over the original)
- It doesn't require us to mandate a particular arr attr format
- The parameter list isn't as indecipherable

On the negatives side, the performance is arguably worse,
but it seems like a reasonable trade-off. You can still use
cslib_array filter as well.

diegoboston commented on scenes/startup.txt:
```diff
@@ -106,7 +105,7 @@
 	*if (total_tests_passed < total_tests_run)
 		*comment print all module logs
-		*gosub_scene cslib_loop repeat "_test_display_templates" "display_log" module_count
+		*gosub_scene cslib_array for_each "module" "_test_display_templates" "display_log"
```
nice!

diegoboston commented on scenes/cslib_menu.txt:
```diff
@@ -111,34 +149,52 @@
 *label _render_choice
+*comment catch 'no selectable options' errors
+*if (page_item_count = 0) and not(allow_cancel)
+	*bug cslib_menu._populate_page: menu construction failed as there were 0 matching items — consider allowing cancellation
```
great

CareyJWilliams committed:
Pass 'array_ref' parameter to argument routines

Allow array routines in cslib to refer
to refer to the array itself via an 'array_ref'
parameter.

This is useful when you want to manipulate
(or check) other attributes associated with
that array as part of the routine's behaviour.
